internal/ent/hooks/event.go

```diff
@@ -0,0 +1,249 @@
+			retVal, err := next.Mutate(ctx, mutation)
+			if err != nil {
+				return nil, err
+			}
+
+			// if the return value is an int, it's a count of the number of rows affected
+			// and we don't want to emit an event for that
+			if reflect.TypeOf(retVal).Kind() == reflect.Int {
+				log.Debug().Interface("value", retVal).Msg("mutation returned an int, skipping event emission")
+
+				return retVal, err
+			}
+
+			op := mutation.Op()
+			typ := mutation.Type()
+			fields := mutation.Fields()
+
+			out, err := json.Marshal(retVal)
+			if err != nil {
+				log.Err(err).Msg("Failed to marshal return value")
+			}
+
+			other := EventID{}
+			if err := json.Unmarshal(out, &other); err != nil {
+				log.Err(err).Msg("Failed to unmarshal return value")
+			}
+
+			// create a new event unfilitered for every mutation
+			// this pushes events into the pool but they are only actioned against if there is a listener
+			event := soiree.NewBaseEvent(fmt.Sprintf("%s.%s", typ, op), mutation)
+			event.Properties().Set("ID", other.ID)
+
+			for _, field := range fields {
+				value, exists := mutation.Field(field)
+				if exists {
+					event.Properties().Set(field, value)
+				}
+			}
+
+			event.SetContext(context.WithoutCancel(ctx))
+			event.SetClient(pool.GetClient())
+			pool.Emit(event.Topic(), event)
+
+			return retVal, err
+		})
+	}
+}
+
+// RegisterGlobalHooks registers global hooks for the ent client with the initialized event pool
+func RegisterGlobalHooks(client *entgen.Client, pool *soiree.EventPool) {
+	client.Use(EmitEventHook(pool))
+}
+
+// InitEventPool initializes an event pool with a client and a error handler
+func InitEventPool(client interface{}) *soiree.EventPool {
+	return soiree.NewEventPool(soiree.WithPool(soiree.NewPondPool(soiree.WithMaxWorkers(100), soiree.WithName("ent_event_pool"))), soiree.WithErrorHandler(func(event soiree.Event, err error) error { // nolint: mnd
+		log.Printf("Error encountered during event '%s': %v, with payload: %v", event.Topic(), err, event.Payload())
+		return nil
+	}), soiree.WithClient(client))
+}
+
+// TODO: template out listeners for all mutation types and create hooks that allow functions to be easily called from within the codebase to perform actions on those events
+// RegisterListeners registers listeners for events on the event pool and is registered with the dbclient
+func RegisterListeners(pool *soiree.EventPool) {
+	_, err := pool.On("OrganizationSetting.OpCreate", handleCustomerCreate)
+	if err != nil {
+		log.Error().Err(ErrFailedToRegisterListener)
+	}
+
+	for _, event := range []string{"OrganizationSetting.OpUpdate", "OrganizationSetting.OpUpdateOne"} {
+		_, err = pool.On(event, handleCustomerCreate)
+		if err != nil {
+			log.Error().Err(ErrFailedToRegisterListener)
+		}
+	}
+}
+
+// handleCustomerCreate handles the creation of a customer in Stripe
+func handleCustomerCreate(event soiree.Event) error {
```

Comment:
this function is pretty complex, I would suggest splitting out the "update stripe record" vs. "create new stripe record" into their own functions that are called from here

Reply:
I'm not sure I can do that as easily as you're suggesting (I've tried to simplify it as much as possible already) without creating more complex data mapping structures. Each of the `Update` commands are using different inputs based on whether or not there is a match because we don't want to stomp on a field that should not be updated (e.g. if a mutation does not contain the field we don't want to overwrite what's already in stripe)